circuits/cpp/barretenberg/cpp/src/barretenberg/proof_system/relations/ecc_vm/ecc_transcript_relation.cpp

```diff
@@ -0,0 +1,263 @@
+    auto transcript_accumulator_x = View(extended_edges.transcript_accumulator_x);
+    auto transcript_accumulator_y = View(extended_edges.transcript_accumulator_y);
+    auto transcript_msm_x = View(extended_edges.transcript_msm_x);
+    auto transcript_msm_y = View(extended_edges.transcript_msm_y);
+    auto transcript_x = View(extended_edges.transcript_x);
+    auto transcript_y = View(extended_edges.transcript_y);
+    auto is_accumulator_empty = View(extended_edges.transcript_accumulator_empty);
+    auto lagrange_first = View(extended_edges.lagrange_first);
+    auto lagrange_last = View(extended_edges.lagrange_last);
+    auto is_accumulator_empty_shift = View(extended_edges.transcript_accumulator_empty_shift);
+    auto q_reset_accumulator = View(extended_edges.transcript_reset_accumulator);
+    auto lagrange_second = View(extended_edges.lagrange_second);
+    auto transcript_collision_check = View(extended_edges.transcript_collision_check);
+
+    auto is_not_first_row = (-lagrange_first + 1);
+    auto is_not_first_or_last_row = (-lagrange_first + -lagrange_last + 1);
+    /**
+     * @brief Validate correctness of z1_zero, z2_zero.
+     * If z1_zero = 0 and operation is a MUL, we will write a scalar mul instruction into our multiplication table.
+     * If z1_zero = 1 and operation is a MUL, we will NOT write a scalar mul instruction.
+     * (same with z2_zero).
+     * z1_zero / z2_zero is user-defined.
+     * We constraint z1_zero such that if z1_zero == 1, we require z1 == 0. (same for z2_zero).
+     * We do *NOT* constrain z1 != 0 if z1_zero = 0. If the user sets z1_zero = 0 and z1 = 0,
+     * this will add a scalar mul instruction into the multiplication table, where the scalar multiplier is 0.
+     * This is inefficient but will still produce the correct output.
+     */
+    std::get<0>(accumulator) += (z1 * z1_zero) * scaling_factor; // if z1_zero = 1, z1 must be 0
+    std::get<1>(accumulator) += (z2 * z2_zero) * scaling_factor;
+
+    /**
+     * @brief Validate `op` opcode is well formed.
+     * `op` is defined to be q_reset_accumulator + 2 * q_eq + 2 * q_mul + 4 * q_add,
+     * where q_reset_accumulator, q_eq, q_mul, q_add are all boolean
+     * (todo, bool constrain?)
+     */
+    auto tmp = q_add + q_add;
+    tmp += q_mul;
+    tmp += tmp;
+    tmp += q_eq;
+    tmp += tmp;
+    tmp += q_reset_accumulator;
+    std::get<2>(accumulator) += (tmp - op) * scaling_factor;
+
+    /**
+     * @brief Validate `pc` is updated correctly.
+     * pc stands for Point Counter. It decrements by 1 for every 128-bit multiplication operation.
+     * If q_mul = 1, pc decrements by !z1_zero + !z2_zero, else pc decrements by 0
+     * @note pc starts out at its max value and decrements down to 0. This keeps the degree of the pc polynomial smol
+     */
+    Accumulator pc_delta = pc - pc_shift;
+    std::get<3>(accumulator) +=
+        is_not_first_row * (pc_delta - q_mul * ((-z1_zero + 1) + (-z2_zero + 1))) * scaling_factor;
```

Comment:
z1_zero and z2_zero are not constrained to be just 1, when z1==0 and z2==0, which can allow a person to decrease pc and msm when z1 and z2 are zero by any number they want